File: modules/supabase-core-python/src/utils/error_handler.py
"""
Error Handler

Utilities for normalizing and handling Supabase errors.
"""

from dataclasses import dataclass
from typing import Optional, Dict, Any

# Common Supabase error codes
SUPABASE_ERROR_CODES: Dict[str, str] = {
    "PGRST116": "No rows returned",
    "23505": "Unique constraint violation",
    "23503": "Foreign key constraint violation",
    "42P01": "Table does not exist",
    "42501": "Insufficient privileges",
    "PGRST301": "JWT expired",
    "PGRST302": "JWT invalid",
}


@dataclass
class NormalizedError:
    """Normalized error structure."""

    message: str
    code: Optional[str] = None
    status_code: Optional[int] = None
    details: Optional[Dict[str, Any]] = None


def normalize_error(error: Any) -> NormalizedError:
    """
    Normalizes a Supabase error into a standard structure.

    Args:
        error: Error object from Supabase

    Returns:
        Normalized error structure

    Example:
        ```python
        from supabase_core_python import normalize_error

        try:
            response = supabase.table("posts").insert(data).execute()
        except Exception as e:
            normalized = normalize_error(e)
            print(f"Error: {normalized.message} (Code: {normalized.code})")
        ```
    """
    # Handle different error types
    if isinstance(error, dict):
        message = error.get("message", str(error))
        code = error.get("code") or error.get("error_code")
        status_code = error.get("status_code") or error.get("status")
        details = error.get("details") or error.get("hint")
    elif hasattr(error, "message"):
        message = str(error.message)
        code = getattr(error, "code", None)
        status_code = getattr(error, "status_code", None)
        details = getattr(error, "details", None)
    else:
        message = str(error)
        code = None
        status_code = None
        details = None

    return NormalizedError(
        message=message,
        code=code,
        status_code=status_code,
        details={"details": details} if details else None,
    )


def is_error_code(error: Any, code: str) -> bool:
    """
    Checks if an error matches a specific error code.

    Args:
        error: Error object
        code: Error code to check for

    Returns:
        True if error code matches
    """
    normalized = normalize_error(error)
    return normalized.code == code


def get_user_friendly_message(error: Any) -> str:
    """
    Gets a user-friendly error message from a Supabase error.

    Args:
        error: Error object

    Returns:
        User-friendly error message

    Example:
        ```python
        from supabase_core_python import get_user_friendly_message

        try:
            response = supabase.table("posts").insert(data).execute()
        except Exception as e:
            user_message = get_user_friendly_message(e)
            return {"error": user_message}
        ```
    """
    normalized = normalize_error(error)

    # Check if we have a friendly message for this error code
    if normalized.code and normalized.code in SUPABASE_ERROR_CODES:
        return SUPABASE_ERROR_CODES[normalized.code]

    # Return the error message
    return normalized.message

File: modules/supabase-core-typescript/src/utils/error-handler.ts
/**
 * Error Handler
 * 
 * Utilities for normalizing and handling Supabase errors consistently.
 * 
 * Dependencies: @supabase/supabase-js
 */

/**
 * Normalized Supabase error.
 */
export interface NormalizedError {
  /** Error code */
  code: string
  /** Human-readable error message */
  message: string
  /** Original error details */
  details?: any
  /** HTTP status code (if applicable) */
  statusCode?: number
}

/**
 * Common Supabase error codes.
 */
export const SUPABASE_ERROR_CODES = {
  // Auth errors
  INVALID_CREDENTIALS: 'invalid_credentials',
  USER_NOT_FOUND: 'user_not_found',
  EMAIL_NOT_CONFIRMED: 'email_not_confirmed',
  TOKEN_EXPIRED: 'token_expired',

  // Database errors
  UNIQUE_VIOLATION: '23505',
  FOREIGN_KEY_VIOLATION: '23503',
  NOT_NULL_VIOLATION: '23502',
  CHECK_VIOLATION: '23514',

  // RLS errors
  RLS_POLICY_VIOLATION: '42501',

  // Storage errors
  STORAGE_NOT_FOUND: 'storage_not_found',
  STORAGE_UNAUTHORIZED: 'storage_unauthorized',
} as const

/**
 * Normalizes a Supabase error into a consistent format.
 * 
 * @param error - Error from Supabase operation
 * @returns Normalized error object
 * 
 * @example
 * ```typescript
 * const { data, error } = await supabase.from('posts').insert({ title: 'New Post' })
 * if (error) {
 *   const normalized = normalizeError(error)
 *   console.error('Error:', normalized.message, normalized.code)
 * }
 * ```
 */
export function normalizeError(error: any): NormalizedError {
  // Handle PostgrestError
  if (error?.code && error?.message) {
    return {
      code: error.code,
      message: error.message,
      details: error.details,
      statusCode: error.status || 500,
    }
  }

  // Handle AuthError
  if (error?.status && error?.message) {
    return {
      code: error.status.toString(),
      message: error.message,
      details: error,
      statusCode: error.status,
    }
  }

  // Handle StorageError
  if (error?.error) {
    return {
      code: error.error || 'storage_error',
      message: error.message || 'Storage operation failed',
      details: error,
    }
  }

  // Handle generic Error
  if (error instanceof Error) {
    return {
      code: 'unknown_error',
      message: error.message,
      details: error,
    }
  }

  // Fallback
  return {
    code: 'unknown_error',
    message: String(error || 'An unknown error occurred'),
    details: error,
  }
}

/**
 * Checks if an error is a specific Supabase error code.
 * 
 * @param error - Error to check
 * @param code - Error code to match
 * @returns True if error matches the code
 * 
 * @example
 * ```typescript
 * if (isErrorCode(error, SUPABASE_ERROR_CODES.UNIQUE_VIOLATION)) {
 *   console.log('Duplicate entry detected')
 * }
 * ```
 */
export function isErrorCode(error: any, code: string): boolean {
  const normalized = normalizeError(error)
  return normalized.code === code
}

/**
 * Gets a user-friendly error message from a Supabase error.
 * 
 * @param error - Error from Supabase operation
 * @returns User-friendly message
 * 
 * @example
 * ```typescript
 * const message = getUserFriendlyMessage(error)
 * // Returns: "This email is already registered" instead of "23505: duplicate key value"
 * ```
 */
export function getUserFriendlyMessage(error: any): string {
  const normalized = normalizeError(error)

  // Map common error codes to user-friendly messages
  const friendlyMessages: Record<string, string> = {
    [SUPABASE_ERROR_CODES.UNIQUE_VIOLATION]: 'This record already exists',
    [SUPABASE_ERROR_CODES.FOREIGN_KEY_VIOLATION]: 'Referenced record does not exist',
    [SUPABASE_ERROR_CODES.NOT_NULL_VIOLATION]: 'Required field is missing',
    [SUPABASE_ERROR_CODES.INVALID_CREDENTIALS]: 'Invalid email or password',
    [SUPABASE_ERROR_CODES.EMAIL_NOT_CONFIRMED]: 'Please verify your email address',
    [SUPABASE_ERROR_CODES.TOKEN_EXPIRED]: 'Your session has expired. Please sign in again',
    [SUPABASE_ERROR_CODES.RLS_POLICY_VIOLATION]: 'You do not have permission to perform this action',
    [SUPABASE_ERROR_CODES.STORAGE_NOT_FOUND]: 'File not found',
    [SUPABASE_ERROR_CODES.STORAGE_UNAUTHORIZED]: 'You do not have permission to access this file',
  }

  return friendlyMessages[normalized.code] || normalized.message
}

